pandas/tests/indexes/test_base.py

```diff
@@ -1034,6 +1034,29 @@ def test_symmetric_difference(self):
         assert tm.equalContents(result, expected)
         assert result.name == 'new_name'
 
+    def test_difference_type(self):
+        # GH 20040
+        # If taking difference of a set and itself, it
+        # needs to preserve the type of the index
+        skip_index_keys = ['repeats']
+        for key, id in self.indices.items():
+            if key not in skip_index_keys:
+                result = id.difference(id)
+                expected = id.drop(id)
+                tm.assert_index_equal(result, expected)
+
+    def test_intersection_difference(self):
+        # GH 20040
+        # Test that the intersection of an index with an
+        # empty index produces the same index as the difference
+        # of an index with itself.  Test for all types
+        skip_index_keys = ['repeats']
```

Comment:
for followup PR (or can change here if want to update where needed).  a function to return yeld the key, idx while filtering on certain types would be great. (like what you are doing but in a module level function)